modules/fedimint-lnv2-client/src/send_sm.rs

```diff
@@ -0,0 +1,338 @@
+        let funded = match old_state.state {
+            SendSMState::Funded(funded) => funded,
+            _ => panic!("Unexpected prior state"),
+        };
+
+        match gateway_response {
+            Ok(preimage) => SendStateMachine {
+                operation_id: old_state.operation_id,
+                contract: old_state.contract,
+                state: SendSMState::Success(preimage),
+            },
+            Err(signature) => {
+                let client_input = ClientInput::<LightningInput, LightningClientStateMachines> {
+                    input: LightningInput {
+                        amount: old_state.contract.amount,
+                        witness: Witness::Outgoing(
+                            old_state.contract.contract_id(),
+                            OutgoingWitness::Cancel(signature),
+                        ),
+                    },
+                    keys: vec![funded.refund_keypair],
+                    // The input of the refund tx is managed by this state machine
+                    state_machines: Arc::new(|_, _| vec![]),
+                };
+
+                let out_points = global_context.claim_input(dbtx, client_input).await.1;
+
+                SendStateMachine {
+                    operation_id: old_state.operation_id,
+                    contract: old_state.contract,
+                    state: SendSMState::Refunding(out_points),
+                }
+            }
+        }
+    }
+
+    async fn await_preimage(
+        global_context: DynGlobalClientContext,
+        contract: OutgoingContract,
+    ) -> Option<[u8; 32]> {
+        loop {
+            let preimage = global_context
+                .module_api()
+                .await_preimage(&contract.contract_id(), contract.expiration)
+                .await;
+
+            match preimage {
+                Some(preimage) => {
+                    if contract.verify_preimage(&preimage) {
+                        return Some(preimage);
+                    }
+
+                    error!("Federation returned invalid preimage {:?}", preimage);
+                }
+                None => return None,
+            }
+
+            sleep(RETRY_DELAY).await;
+        }
+    }
+
+    async fn transition_preimage(
+        dbtx: &mut ClientSMDatabaseTransaction<'_, '_>,
+        global_context: DynGlobalClientContext,
+        old_state: SendStateMachine,
+        preimage: Option<[u8; 32]>,
+    ) -> SendStateMachine {
+        if let Some(preimage) = preimage {
+            return SendStateMachine {
+                operation_id: old_state.operation_id,
+                contract: old_state.contract,
+                state: SendSMState::Success(preimage),
+            };
+        }
+
+        let funded = match old_state.state {
+            SendSMState::Funded(funded) => funded,
+            _ => panic!("Unexpected prior state"),
+        };
+
+        let client_input = ClientInput::<LightningInput, LightningClientStateMachines> {
```

Comment:
Perhaps make this and the logic below into a function, I see it is used above when the gateway returns a forfeit signature as well.